cuDF error handling document (#7917)
rapidsai/cudf: docs/cudf/source/developer_guide/errors.md

# cuDF Error Handling Documentation

This document discusses what to raise given invalid user inputs and how to handle libcudf exceptions in cuDF.

## Overview

> cuDF uses [builtin error types](https://docs.python.org/3/library/exceptions.html) to indicate user error.

Reviewer: Is this intended to imply that cuDF should not use custom exception types?

Author: I don't think that's intended. How about
```python
cuDF uses [builtin error types](https://docs.python.org/3/library/exceptions.html) and custom error types to indicate user error.
```

Reviewer: I don't think this sentence really adds anything. I would just remove it.

If an API shares an argument with pandas,

> cuDf should raise the same error type given a specific invalid input for that argument.

Reviewer: ```suggestion
cuDF should raise the same error type given a specific invalid input for that argument.
```

> Developer of the API should include sufficient information in the exception message to helper user locate the source of the error,

Reviewer: ```suggestion
Developers should include sufficient information in the exception message to help users locate the source of the error,
```

> but it is not required to match pandas API's exception message.

Reviewer: ```suggestion
but it is not required to match the corresponding pandas API's exception message.
```

For parameters that are not yet supported,
raise [`NotImplementedError`](https://docs.python.org/3/library/exceptions.html#NotImplementedError).

## Handling libcudf Exceptions

> Currently all libcudf expcetions are `cudf::logic_error` type which inherits `std::logic_error`.

Reviewer: ```suggestion
Currently all libcudf exceptions are of type `cudf::logic_error`, which inherits from `std::logic_error`.
```

Reviewer: This isn't quite true, libcudf can also throw a `cuda_error`.

> By default the error type is mapped to `RuntimeError` in python.

Reviewer: ```suggestion
By default this error type is mapped to `RuntimeError` in python.
```
